File: mac/Sources/FocusKit/Models/Note.swift
import Foundation

public struct NoteColor: Identifiable, Equatable, Sendable {
    public let key: String
    public let label: String
    public let rgb: (red: Double, green: Double, blue: Double)

    public var id: String { key }

    public static func == (lhs: NoteColor, rhs: NoteColor) -> Bool { lhs.key == rhs.key }

    public static let all: [NoteColor] = [
        NoteColor(key: "amber", label: "Amber", rgb: (1.00, 0.83, 0.37)),
        NoteColor(key: "pink", label: "Pink", rgb: (1.00, 0.62, 0.77)),
        NoteColor(key: "mint", label: "Mint", rgb: (0.56, 0.89, 0.75)),
        NoteColor(key: "sky", label: "Sky", rgb: (0.56, 0.80, 1.00)),
        NoteColor(key: "lilac", label: "Lilac", rgb: (0.76, 0.65, 1.00)),
        NoteColor(key: "lime", label: "Lime", rgb: (0.85, 0.91, 0.42))
    ]

    public static func named(_ key: String) -> NoteColor {
        all.first { $0.key == key } ?? all[0]
    }
}

/// A sticky note. Its text lives in its own Word document on disk; everything here
/// but `text` is board furniture, kept in board.json beside it.
public struct Note: Identifiable, Equatable {
    public static let defaultSize = CGSize(width: 240, height: 240)
    public static let minimumSize = CGSize(width: 160, height: 150)

    /// The note's filename, which is also its identity on disk.
    public let id: String
    public let createdAt: Date
    public var text: String
    public var x: Double
    public var y: Double
    public var width: Double
    public var height: Double
    public var color: String
    public var z: Int

    public var size: CGSize { CGSize(width: width, height: height) }

    public var title: String {
        let first = text.split(separator: "\n", omittingEmptySubsequences: false).first.map(String.init) ?? ""
        let trimmed = first.trimmingCharacters(in: .whitespaces)
        return trimmed.isEmpty ? "Untitled note" : trimmed
    }

    public init(
        id: String,
        createdAt: Date,
        text: String = "",
        x: Double = 24,
        y: Double = 24,
        width: Double = Note.defaultSize.width,
        height: Double = Note.defaultSize.height,
        color: String = "amber",
        z: Int = 1
    ) {
        self.id = id
        self.createdAt = createdAt
        self.text = text
        self.x = x
        self.y = y
        self.width = width
        self.height = height
        self.color = color
        self.z = z
    }
}

/// What board.json holds: where a note sits, not what it says.
struct NoteLayout: Codable {
    var x: Double
    var y: Double
    var width: Double
    var height: Double
    var color: String
    var z: Int
}

File: windows/Focus.Core/Models/Note.cs
namespace Focus.Core;

public sealed record NoteColor(string Key, string Label, string Hex)
{
    public static readonly NoteColor[] All =
    {
        new("amber", "Amber", "#FFD45E"),
        new("pink", "Pink", "#FF9EC4"),
        new("mint", "Mint", "#8EE3BF"),
        new("sky", "Sky", "#8FCCFF"),
        new("lilac", "Lilac", "#C2A6FF"),
        new("lime", "Lime", "#D9E86B")
    };

    public static NoteColor Named(string key) =>
        All.FirstOrDefault(c => c.Key == key) ?? All[0];
}

/// A sticky note. Its text is a section of the board's Word document; everything
/// else here is board furniture, kept in Board.json beside it.
public sealed class Note : Observable
{
    public const double DefaultWidth = 240;
    public const double DefaultHeight = 240;
    public const double MinimumWidth = 160;
    public const double MinimumHeight = 150;

    private string _text = "";
    private double _x = 24, _y = 24, _width = DefaultWidth, _height = DefaultHeight;
    private string _color = "amber";
    private int _z = 1;

    /// The note's stamp, which is also its identity inside the document.
    public string Id { get; }
    public DateTime CreatedAt { get; }

    public Note(string id, DateTime createdAt) { Id = id; CreatedAt = createdAt; }

    public string Text { get => _text; set { if (Set(ref _text, value)) Raise(nameof(Title)); } }
    public double X { get => _x; set => Set(ref _x, value); }
    public double Y { get => _y; set => Set(ref _y, value); }
    public double Width { get => _width; set => Set(ref _width, value); }
    public double Height { get => _height; set => Set(ref _height, value); }
    public string Color { get => _color; set => Set(ref _color, value); }
    public int Z { get => _z; set => Set(ref _z, value); }

    public string Title
    {
        get
        {
            var first = (Text.Split('\n').FirstOrDefault() ?? "").Trim();
            return first.Length == 0 ? "Untitled note" : first;
        }
    }
}

/// What Board.json holds: where a note sits, not what it says.
public sealed class NoteLayout
{
    public double X { get; set; }
    public double Y { get; set; }
    public double Width { get; set; }
    public double Height { get; set; }
    public string Color { get; set; } = "amber";
    public int Z { get; set; }
}
